### DISCUSSION TITLE
Reproducibility of predict and cluster_pairwise_predictions_at_threshold

### DISCUSSION THREAD

**Discussion by sthamodh** [moj-analytical-services/splink#1850] [Q&A]
Hi there,

I have been experimenting with a few hyperparameters for a given (a singular input dataset for deduplication and I use SparkLinker) splink run such as:

1. probability_two_random_records_match
2. em_convergence
3. max_pairs_for_estimating_u

Let us say I have found a combination that has the least value for a certain metric such as levenshtein distance of left and right records after splink's has produced an output table using the linker object's predict function.

After finding such a value, I am unable to reproduce the same results when I plug in these optimal values. This also affects the number of records in the pairwise records table as well as the cluster table.

How would I go about replicating a run (with the same exact duplicate sets from the predict and the same exact cluster records in from the cluster table) when I am able fix the three hyperparams above?

1. The only seed parameter I can see is the seed parameter in estimate_u_using_random_sampling function. This has been set to the value 42
2. I have also imported random and set a seed using random.seed(42)

Thanks,

**Comment by RobinL** [comment]
Hmm - not sure.  Are you able to write a reproducible example that gives different answers?  Splink _should_ give the same answer every time if you set the seed, but there may be a bug

**Reply by sthamodh** [reply]
Thanks @RobinL, do you think my post is tied to this issue: https://github.com/moj-analytical-services/splink/issues/1743 ?

As @James-Osmond outlines in the issue, my code also uses a more complex comparison list from the comparison levels library.

If the same m- and u- probabilities are not obtained when setting a seed, would this mean that slightly different pairwise duplicates and clusters are generated as output tables?

**Reply by RobinL** [reply]
Yes - i think it's likely the same issue - good digging.  And yes, there is slight random sampling variability in the estimates of the m and u probabilities (the 'u's specifically), and this could result in slightly different dupes/clusters.  One mitigation would be setting max_pairs higher (e.g. 1e7 or 1e8) [here](https://moj-analytical-services.github.io/splink/linker.html?h=max_pairs#splink.linker.Linker.estimate_u_using_random_sampling).  Start low and increase to gauge the effect on runtime, and I recommend making sure you have splink>=3.9.11 because that will ensure the calculation is fully parallelised.
